spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/query/Criteria.java

```diff
@@ -989,6 +991,34 @@ private boolean isEqual(Object left, Object right) {
 					&& leftPattern.flags() == rightPattern.flags();
 		}
 
+		if (left instanceof Document) {
+			if (!(right instanceof Document)) {
+				return false;
+			}
+			Document leftDocument = (Document) left;
+			Document rightDocument = (Document) right;
+
+			return isEqual(leftDocument.values(), rightDocument.values());
```

Comment:
An equality check for `Map`-like types requires a comparison of keys and values. Right now, a document could yield equality if only the order of values is the same regardless of the key the values is assigned to.

Reply:
Sorry, I totally missed that one. I fixed it now: the keys and the values are compared in the case of a `Document`.